tests/compose-spec_test.go

```diff
@@ -0,0 +1,227 @@
+            "49100:22",
+            "8000-9000:80",
+            "127.0.0.1:8001:8001",
+            "127.0.0.1:5000-5010:5000-5010",
+            "6060:6060/udp"
+        ]`),
+		},
+		{
+			name: "ports cannot be empty",
+			spec: createComposeSpecService(`
+        "ports": [
+            ""
+        ]`),
+			assertError: true,
+		},
+		{
+			name: "ports cannot be negative",
+			spec: createComposeSpecService(`
+        "ports": [
+            "-3000"
+        ]`),
+			assertError: true,
+		},
+		{
+			name: "ports cannot be imcomplete",
+			spec: createComposeSpecService(`
+        "ports": [
+            ":80"
+        ]`),
+			assertError: true,
+		},
+		{
+			name: "ports must have valid protocol",
+			spec: createComposeSpecService(`
+        "ports": [
+          "6060:6060/random"
+        ]`),
+			assertError: true,
+		},
+		{
+			name: "stop_grace_period must be a valid duration (micro second)",
+			spec: createComposeSpecService(`"stop_grace_period": "10us"`),
+		},
+		{
+			name: "stop_grace_period must be a valid duration (milisecond)",
+			spec: createComposeSpecService(`"stop_grace_period": "10ms"`),
+		},
+		{
+			name: "stop_grace_period must be a valid duration (second)",
+			spec: createComposeSpecService(`"stop_grace_period": "10s"`),
+		},
+		{
+			name: "stop_grace_period must be a valid duration (minute)",
+			spec: createComposeSpecService(`"stop_grace_period": "10m"`),
+		},
+		{
+			name: "stop_grace_period must be a valid duration (hour)",
+			spec: createComposeSpecService(`"stop_grace_period": "10h"`),
+		},
+		{
+			name: "stop_grace_period must be a valid duration (combined)",
+			spec: createComposeSpecService(`"stop_grace_period": "10000h1000m100s10ms1us"`),
+		},
+		{
+			name:        "stop_grace_period cannot be given a duration with wrong order",
+			spec:        createComposeSpecService(`"stop_grace_period": "1s10h"`),
+			assertError: true,
+		},
+		{
+			name:        "stop_grace_period cannot be given a duration with wrong unit",
+			spec:        createComposeSpecService(`"stop_grace_period": "1kg"`),
+			assertError: true,
+		},
+		{
+			name: "deploy/rollback_config/delay is a duration",
+			spec: createComposeSpecService(`"deploy": {"rollback_config": {"delay": "1h2s3ms4us"}}`),
+		},
+		{
+			name: "deploy/rollback_config/monitor is a duration",
+			spec: createComposeSpecService(`"deploy": {"rollback_config": {"monitor": "1h2s3ms4us5ns"}}`),
+		},
+		{
+			name: "optional x- should be allowed at top level",
+			spec: createComposeSpec(`"x-test":"something"`),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			sch, _ := jsonschema.Compile("../schema/compose-spec.json")
+
+			var data interface{}
+
+			// NOTE: make sure in array test element case does not have a ',' in the last element
+			err := json.Unmarshal([]byte(tt.spec), &data)
+			assert.NilError(t, err)
+
+			err = sch.Validate(data)
+			if !tt.assertError {
+				assert.NilError(t, err)
```

Comment:
You can avoid the `gotest.tools/v3` dependency by doing direct calls to `t.Fatalf` here (like the "should have thrown error for spec" assertion)